File: app/src/main/java/com/mappls/sdk/demo/java/plugin/SnakePolyLinePlugin.java
package com.mappls.sdk.demo.java.plugin;

import static com.mappls.sdk.maps.style.layers.Property.LINE_CAP_ROUND;
import static com.mappls.sdk.maps.style.layers.Property.LINE_JOIN_ROUND;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineCap;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineColor;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineJoin;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineOpacity;
import static com.mappls.sdk.maps.style.layers.PropertyFactory.lineWidth;

import android.graphics.Color;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;

import com.mappls.sdk.geojson.Feature;
import com.mappls.sdk.geojson.FeatureCollection;
import com.mappls.sdk.geojson.LineString;
import com.mappls.sdk.maps.MapView;
import com.mappls.sdk.maps.MapplsMap;
import com.mappls.sdk.maps.Style;
import com.mappls.sdk.maps.style.layers.LineLayer;
import com.mappls.sdk.maps.style.sources.GeoJsonSource;
import com.mappls.sdk.services.api.directions.models.LegStep;
import com.mappls.sdk.services.utils.Constants;

import java.util.ArrayList;
import java.util.List;


public class SnakePolyLinePlugin implements MapView.OnDidFinishLoadingStyleListener {

    private MapView mapView;
    private MapplsMap mapplsMap;
    private static final float NAVIGATION_LINE_WIDTH = 6;
    private static final float NAVIGATION_LINE_OPACITY = .8f;
    private static final String DRIVING_ROUTE_POLYLINE_LINE_LAYER_ID = "DRIVING_ROUTE_POLYLINE_LINE_LAYER_ID";
    private static final String DRIVING_ROUTE_POLYLINE_SOURCE_ID = "DRIVING_ROUTE_POLYLINE_SOURCE_ID";
    private static final int DRAW_SPEED_MILLISECONDS = 500;
    private Handler handler = new Handler(Looper.getMainLooper());
    private List<LegStep> legSteps;
    private Runnable runnable;

    public SnakePolyLinePlugin(MapView mapView, MapplsMap mapplsMap) {
        this.mapView = mapView;
        this.mapplsMap = mapplsMap;
        mapView.addOnDidFinishLoadingStyleListener(this);
        initialiseSourceAndLayer();
    }

    private void initialiseSourceAndLayer() {
        mapplsMap.getStyle(new Style.OnStyleLoaded() {
            @Override
            public void onStyleLoaded(@NonNull Style style) {
                addSource(style);
                addLayer(style);
            }
        });

    }

    private void addLayer(Style style) {
        if(style.getLayer(DRIVING_ROUTE_POLYLINE_LINE_LAYER_ID) == null) {
            style.addLayer(new LineLayer(DRIVING_ROUTE_POLYLINE_LINE_LAYER_ID,
                    DRIVING_ROUTE_POLYLINE_SOURCE_ID)
                    .withProperties(
                            lineWidth(NAVIGATION_LINE_WIDTH),
                            lineOpacity(NAVIGATION_LINE_OPACITY),
                            lineCap(LINE_CAP_ROUND),
                            lineJoin(LINE_JOIN_ROUND),
                            lineColor(Color.BLUE)
                    ));
        }
    }


    private void addSource(Style style) {
        if(style.getSource(DRIVING_ROUTE_POLYLINE_SOURCE_ID) == null) {
            style.addSource(new GeoJsonSource(DRIVING_ROUTE_POLYLINE_SOURCE_ID));
        }
    }


    public void create(List<LegStep> legSteps) {
        this.legSteps = legSteps;
        // Start the step-by-step process of drawing the route
        runnable = new DrawRouteRunnable(mapplsMap,legSteps, handler);
        handler.postDelayed(runnable, DRAW_SPEED_MILLISECONDS);
    }

    public void removeCallback() {
        if(runnable != null) {
            handler.removeCallbacks(runnable);
        }
    }



    @Override
    public void onDidFinishLoadingStyle() {
        handler.removeCallbacks(runnable);
        runnable = new DrawRouteRunnable(mapplsMap,legSteps, handler);
        handler.postDelayed(runnable, DRAW_SPEED_MILLISECONDS);
    }

    private static class DrawRouteRunnable implements Runnable {
        private MapplsMap mapplsMap;
        private List<LegStep> steps;
        private List<Feature> drivingRoutePolyLineFeatureList;
        private Handler handler;
        private int counterIndex;

        DrawRouteRunnable(MapplsMap mapplsMap, List<LegStep> steps, Handler handler) {
            this.mapplsMap = mapplsMap;
            this.steps = steps;
            this.handler = handler;
            this.counterIndex = 0;
            drivingRoutePolyLineFeatureList = new ArrayList<>();
        }

        @Override
        public void run() {
            if(steps==null)
                return;
            if (  counterIndex < steps.size()) {
                LegStep singleStep = steps.get(counterIndex);
                if (singleStep != null && singleStep.geometry() != null) {
                    LineString lineStringRepresentingSingleStep = LineString.fromPolyline(
                            singleStep.geometry(), Constants.PRECISION_6);
                    Feature featureLineString = Feature.fromGeometry(lineStringRepresentingSingleStep);
                    drivingRoutePolyLineFeatureList.add(featureLineString);
                }
                if (mapplsMap != null) {
                    mapplsMap.getStyle(new Style.OnStyleLoaded() {
                        @Override
                        public void onStyleLoaded(@NonNull Style style) {
                            GeoJsonSource source = style.getSourceAs(DRIVING_ROUTE_POLYLINE_SOURCE_ID);
                            if (source != null) {
                                source.setGeoJson(FeatureCollection.fromFeatures(drivingRoutePolyLineFeatureList));
                            }
                        }
                    });

                }
                counterIndex++;
                handler.postDelayed(this, DRAW_SPEED_MILLISECONDS);
            }
        }
    }
}

File: app/src/main/java/com/mappls/sdk/demo/plugin/SnakePolyLinePlugin.kt
package com.mappls.sdk.demo.plugin

import android.graphics.Color
import com.mappls.sdk.geojson.Feature
import com.mappls.sdk.geojson.FeatureCollection
import com.mappls.sdk.geojson.LineString
import com.mappls.sdk.maps.MapView
import com.mappls.sdk.maps.MapplsMap
import com.mappls.sdk.maps.Style
import com.mappls.sdk.maps.style.layers.LineLayer
import com.mappls.sdk.maps.style.layers.Property
import com.mappls.sdk.maps.style.layers.Property.LINE_CAP_ROUND
import com.mappls.sdk.maps.style.layers.Property.LINE_JOIN_ROUND
import com.mappls.sdk.maps.style.layers.PropertyFactory.lineCap
import com.mappls.sdk.maps.style.layers.PropertyFactory.lineColor
import com.mappls.sdk.maps.style.layers.PropertyFactory.lineJoin
import com.mappls.sdk.maps.style.layers.PropertyFactory.lineOpacity
import com.mappls.sdk.maps.style.layers.PropertyFactory.lineWidth
import com.mappls.sdk.maps.style.sources.GeoJsonSource
import com.mappls.sdk.services.api.directions.models.LegStep
import com.mappls.sdk.services.utils.Constants
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch


class SnakePolyLinePlugin(
    private val mapView: MapView,
    private val mapplsMap: MapplsMap
) : MapView.OnDidFinishLoadingStyleListener {

    companion object {
        private const val NAVIGATION_LINE_WIDTH = 6f
        private const val NAVIGATION_LINE_OPACITY = 0.8f
        private const val DRIVING_ROUTE_POLYLINE_LINE_LAYER_ID = "DRIVING_ROUTE_POLYLINE_LINE_LAYER_ID"
        private const val DRIVING_ROUTE_POLYLINE_SOURCE_ID = "DRIVING_ROUTE_POLYLINE_SOURCE_ID"
        private const val DRAW_SPEED_MILLISECONDS = 500L
    }

    private var legSteps: List<LegStep>? = null
    private var drawJob: Job? = null
    private val coroutineScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    init {
        mapView.addOnDidFinishLoadingStyleListener(this)
        initialiseSourceAndLayer()
    }

    private fun initialiseSourceAndLayer() {
        mapplsMap.getStyle { style ->
            addSource(style)
            addLayer(style)
        }
    }

    private fun addLayer(style: Style) {
        if (style.getLayer(DRIVING_ROUTE_POLYLINE_LINE_LAYER_ID) == null) {
            style.addLayer(
                LineLayer(DRIVING_ROUTE_POLYLINE_LINE_LAYER_ID, DRIVING_ROUTE_POLYLINE_SOURCE_ID)
                    .withProperties(
                        lineWidth(NAVIGATION_LINE_WIDTH),
                        lineOpacity(NAVIGATION_LINE_OPACITY),
                        lineCap(Property.LINE_CAP_ROUND),
                        lineJoin(Property.LINE_JOIN_ROUND),
                        lineColor(Color.BLUE)
                    )
            )
        }
    }

    private fun addSource(style: Style) {
        if (style.getSource(DRIVING_ROUTE_POLYLINE_SOURCE_ID) == null) {
            style.addSource(GeoJsonSource(DRIVING_ROUTE_POLYLINE_SOURCE_ID))
        }
    }

    fun create(legSteps: List<LegStep>) {
        this.legSteps = legSteps
        startDrawingRoute()
    }

    fun removeCallback() {
        drawJob?.cancel()
    }

    override fun onDidFinishLoadingStyle() {
        removeCallback()
        startDrawingRoute()
    }

    private fun startDrawingRoute() {
        val steps = legSteps ?: return
        drawJob = coroutineScope.launch {
            val featureList = mutableListOf<Feature>()
            for (step in steps) {
                step.geometry()?.let { geometry ->
                    val lineString = LineString.fromPolyline(geometry, Constants.PRECISION_6)
                    featureList.add(Feature.fromGeometry(lineString))
                }

                mapplsMap.getStyle { style ->
                    val source = style.getSourceAs<GeoJsonSource>(DRIVING_ROUTE_POLYLINE_SOURCE_ID)
                    source?.setGeoJson(FeatureCollection.fromFeatures(featureList))
                }

                delay(DRAW_SPEED_MILLISECONDS)
            }
        }
    }
}